sklearn/tree/_export.py

```diff
@@ -986,7 +995,15 @@ def export_text(
     check_is_fitted(decision_tree)
     tree_ = decision_tree.tree_
     if is_classifier(decision_tree):
-        class_names = decision_tree.classes_
+        if class_names is not None and len(class_names) != len(decision_tree.classes_):
+            raise ValueError(
+                "When `class_names` is not None, it should be a list containing as"
+                " many items as `decision_tree.classes_`. Got"
+                f" {len(class_names)} while the tree was fitted with"
+                f" {len(decision_tree.classes_)} classes."
+            )
+        elif class_names is None:
+            class_names = decision_tree.classes_
```

Comment:
Nit: This order for checking is a little clearer to me:

```suggestion
        if class_names is None:
            class_names = decision_tree.classes_
        elif len(class_names) != len(decision_tree.classes_):
            raise ValueError(
                "When `class_names` is a list, it should contain as"
                " many items as `decision_tree.classes_`. Got"
                f" {len(class_names)} while the tree was fitted with"
                f" {len(decision_tree.classes_)} classes."
            )          
```

Also, I think it is less wordy to directly state that "When `class_names` is a list". (The test needs to be updated as well)